server/src/main/java/org/apache/cassandra/sidecar/restore/RingTopologyRefresher.java

```diff
@@ -104,6 +153,86 @@ public TokenRangeReplicasResponse cachedReplicaByTokenRange(RestoreJob restoreJo
+        {
+            synchronized (this)
+            {
+                topology = prepareAndFetch((storageOperations, nodeSettings) -> {
+                    String partitioner = nodeSettings.partitioner();
+                    return replicaByTokenRangePerKeyspace.loadOne(keyspace, k -> storageOperations.tokenRangeReplicas(new Name(keyspace), partitioner));
+                });
+            }
+        }
+        else // get the cached value
+        {
+            topology = replicaByTokenRangePerKeyspace.topologyOfKeyspace(keyspace);
+        }
+
+        return calculateLocalTokenRanges(metadataFetcher, topology);
+    }
+
+    // todo: refactor to a utility class _when_ refactoring TokenRangeReplicasResponse data structure (separate out server and http data representations)
+    @NotNull
+    public static Map<Integer, Set<TokenRange>> calculateLocalTokenRanges(InstanceMetadataFetcher metadataFetcher, TokenRangeReplicasResponse topology)
+    {
+        if (topology == null)
+        {
+            return Collections.emptyMap();
+        }
+
+        // todo: this assumes one C* node per IP address
+        Map<String, Integer> allNodes = topology.replicaMetadata().values().stream()
+                                                .collect(Collectors.toMap(TokenRangeReplicasResponse.ReplicaMetadata::address,
+                                                                          TokenRangeReplicasResponse.ReplicaMetadata::port));
+
+        List<InstanceMetadata> localNodes = metadataFetcher.allLocalInstances();
+        Map<String, InstanceMetadata> localEndpointsToMetadata = new HashMap<>(localNodes.size());
+        for (InstanceMetadata instanceMetadata : localNodes)
+        {
+            populateEndpointToMetadata(instanceMetadata, allNodes, localEndpointsToMetadata);
+        }
+
+        Map<Integer, Set<TokenRange>> localTokenRanges = new HashMap<>(localEndpointsToMetadata.size());
+        for (TokenRangeReplicasResponse.ReplicaInfo ri : topology.writeReplicas())
+        {
+            TokenRange range = new TokenRange(Token.from(ri.start()), Token.from(ri.end()));
+            for (List<String> instanceOfDc : ri.replicasByDatacenter().values())
+            {
+                for (String instanceEndpoint : instanceOfDc)
+                {
+                    // skip the non-local nodes
+                    if (!localEndpointsToMetadata.containsKey(instanceEndpoint))
+                    {
+                        continue;
+                    }
+
+                    InstanceMetadata instanceMetadata = localEndpointsToMetadata.get(instanceEndpoint);
+                    localTokenRanges.compute(instanceMetadata.id(), (key, value) -> {
+                        if (value == null)
+                        {
+                            value = new HashSet<>();
+                        }
+                        value.add(range);
+                        return value;
+                    });
```

Comment:
I think this reads better:
```suggestion
                    InstanceMetadata instanceMetadata = localEndpointsToMetadata.get(instanceEndpoint);
                    if (instanceMetadata == null)
                    {
                        continue;
                    }
                    localTokenRanges.computeIfAbsent(instanceMetadata.id(), k -> new HashSet<>())
                                    .add(range);
```

Reply:
Agreed.